port CAVs/ACC features

calbaker opened the pull request:
https://github.nrel.gov/MBAP/fastsim/issues/380

Commit:
added todo

Commit:
Add to_elements() method.

The to_elements() method transforms a Cycle into a vector of CycleElement for easy processing for microtrip splitting and the like.

Commit:
Add to_microtrips method.

This method splits a drive cycle into microtrips and returns them as a vector.

Commit:
Add more tests and fix issue with time not being corrected on last microtrip.

Commit:
Refactor to_microtrips() to use to_elements().

Commit:
Merge branch 'fastsim-3' into f3/acc-features

Commit:
Add constant jerk trajectory utils.

Adding a new module called cycle_manipulation_utils.

Commit:
Merge branch 'fastsim-3' into f3/acc-features

Commit:
Add documentation strings to ConstantJerkTrajectory.

Commit:
Move cycle manipulation utilities into a submodule.

Commit:
Add docstring to to_microtrips method.

Commit:
Remove unused import from fastsim_2.

Commit:
Add average_step_speeds function.

Commit:
Add docstring for average_step_speeds.

Commit:
Add average_step_speed_at and reorganize code.

Commit:
Add trapz_step_distances with tests.

Commit:
Add trapz_step_start_distance method, function, and tests.

Commit:
Add trapz_distance_for_step fn, method, and test.

Commit:
Add trapz_distance_over_range fn, method, and test.

Commit:
Add time_spent_moving fn, method, and test.

Commit:
Add docstring for fn/method

Commit:
Add create_dist_and_target_speeds_by_microtrip method and test.

Commit:
Add function version of create_distance_and_target_speeds_by_microtrip + tests.

Commit:
Add comment showing direction.

Commit:
Merge branch 'fastsim-3' into f3/acc-features

Commit:
Update pragmas to match updates from main fastsim-3 branch.

Commit:
Add extend_time method to Cycle + tests.

Also add helper function.

Commit:
Add PassingInfo struct and routine + test.

Commit:
Address ruff warning to move isort under the tool.ruff.lint.isort category.

Commit:
Address ruff warning to add module documentation.

Commit:
Turn demos directory into a package to enable running and addressing items via imports.

Commit:
Add initial version of demo_cavs.py and enable getters on Cycle data.

Commit:
Remove magic number to variable

Commit:
Expose a few more methods to Python for drive_cycle.

Commit:
Add interp via ninterp.

Commit:
Merge branch 'fastsim-3' of github.com:NREL/fastsim into f3/acc-features

calbaker commented:
@michael-okeefe , I just merged `fastsim-3` into this branch so be sure to pull.

Commit:
Merge branch 'f3/acc-features' of https://github.com/NREL/fastsim into f3/acc-features

Commit:
Add missing doc string

Commit:
Fix issue with interpolating elevations.

Also adds trapz_step_elevations(), cycle cache builder, average_grade_over_range calculation, and additional unit tests.

Commit:
Apply rust analyzer hints.

Commit:
Update pyo3 to address uses_conversion warning.

See:
https://stackoverflow.com/a/79469914

Commit:
Apply clippy fixes / suggestions.

Commit:
Rename variable for clarity.

Using cc for CycleCache. Used to be RustCycleCache so rcc...

Commit:
Add calc_distance_to_next_stop_from() plus tests.

Commit:
Add modify_by_const_jerk_trajectory + tests.

Note: this commit also adds units and quanity
for Jerk (3rd derivative of position with respect
to time) and jerk units (m/s3).

Commit:
Add modify_with_braking_trajectory plus tests.

Commit:
Begin adding coasting parameters

Commit:
Merge branch 'fastsim-3' into f3/acc-features

Commit:
Add basic coasting parameters.

Also allow setting coasting parameters from Python.

Commit:
Merge branch 'fastsim-3' into f3/acc-features

Commit:
Add reference cycle for CAVs maneuvers.

Commit:
Add coasting variables and support functions

Begins adding in core eco-coasting functions. Still not hooked up to trigger coast yet as of this commit.

Commit:
Enable simple coasting based on speed trigger.

This example demonstrates a coast that is triggered by crossing a speed threshold. Once that speed is crossed, the vehicle coasts to stop and then tries to resume the normal cycle.

Commit:
Update name for clarity.

Commit:
Enhance note for further clarity.

Commit:
Protect from initialization error with default

Commit:
Change to assert to simplify function type.

Commit:
Add maneuver module.

Commit:
Enable maneuver demo to call methods from Rust

Commit:
Fix bug in eco-coast maneuver.

The following bugs were found and fixed:

- Air density was not being set for some reason
- Grade is being re-calculated each step
- Fix using the wrong field from cyc0_cache => was using trapz_step_distances_m (distance traveled each step) but needed to use trapz_distances_m (the distance from start at the beginning of each step)
- update unit test to test that we deviate from original trace

Commit:
Remove eco* params from simdrive and vehicle

The eco-coasting and eco-cruise maneuvers are now in the maneuvers module nested under drive_cycle. As such, we are removing preliminary work to add them directly to simdrive and vehicle as these manuvers can be pre-processed.

Commit:
Merge branch 'fastsim-3' into f3/acc-features

Commit:
Expose extend_time method and add to demo.

The extend_time method allows for a nice way to
extend the duration of an existing cycle. This
is great for eco-coast which (when used alone)
often arrives at the distance target later than
the reference cycle. The demo shows its typical
usage.

Commit:
Annotate demo_cavs.py file for coasting more

Commit:
Remove unnecessary print statement.

Commit:
Update cavs demo.

Commit:
Change modify_with_braking_trajectory signature.

No longer return an anyhow::Result<(...)> but just the tuple to avoid the need for callers to have to unwrap() etc.

Commit:
Fix difference with FASTSim 2.

Commit:
Add assertions to prevent misuse.

Also reference the original n to avoid casting twice.

Commit:
Code cleanup and add additional unit test.

Commit:
Update should_impose_coast to ref previous speed.

For using coast_start_speed, we need to reference the previous achieved speed or else we enter into a cavitation situation.

Commit:
Remove duplicate py getters.

This data can be accessed via to_pydict(.).

Also, added a TODO note.

Commit:
Begin adding eco-cruise parameters and methods.

Commit:
Fix issue with CAVs demo.

Had removed one of the python convenience
methods as the same data can be accessed using
to_pydict. Updated the code to reflect the new
interface.

Commit:
Add basic demo for eco-cruise

Commit:
Add eco-cruise test.

This test just checks that the procedure runs but doesn't check the output.

Commit:
Add convenience methods and expose to python.

Adds tests that check the amount of idle time at cycle end as well as the ability to trim a cycle to include only that amount of idle time. This helps clean up a cycle that has been extended.

Commit:
Update demo to trim cycles

Commit:
Add cruise and coast demo

Commit:
Add cruise+coast test and expose more to py

Expose a method that summarizes whether coasting mode is on for analysis afterwards.

Commit:
Update plots and add coasting mode designation

Commit:
Update idm params to use units of measure.

michael-okeefe commented:
@calbaker Hi Chad, at this point, the main tasks are complete and all tests are passing. We can go over this at our meeting tomorrow but would like to put this forward for merging. For some reason, I was unable to add you as a reviewer but I added @kylecarow . The only remaining items are testing with alternative powertrains (since cycle manipulation occurs outside of simdrive now, I don't expect an issue but we should confirm) and testing with grade (should be working and hooked up but untested), and, finally, integrating an eco-cruise functionality to compensate for trace miss (which could be on a separate smaller branch). Those changes could come in on a separate branch.

(CC @jakeholden )

Commit:
Add docstrings to Rust methods and demo files

Commit:
Merge branch 'fastsim-3' into f3/acc-features

NOTE: conflict in drive_cycle.rs resolved by
adding the unit tests in both pathways.

Commit:
Fix issues caused by change in serdeapi

michael-okeefe commented:
@calbaker @jakeholden
Just updated this branch to add some docstrings and merge fastsim-3 in and address merge conflicts. As soon as the tests pass, this should be ready to merge into the fastsim-3 branch. I'll continue the remaining tasks on another branch and start a PR.

Commit:
Force routine to go into while loop at least once.

We were having issues when starting from 0 speed with solver not solving. This fixes the problem.

Commit:
Add trace miss using constant jerk correction.

This enables us to re-rendezvous with the target trace using the "gentlest" trajectory within the max_step range.

Commit:
Add trace miss options to system and test via demo.

This adds the maximum number of steps to consider for re-rendezvous with the reference trace. The trajectory with the "gentlest" acceleration profile (minimum peak acceleration) is chosen.

Commit:
Add note and time/distance plot.

The time/distance plot is helpful to see where two traces are relative to each other.

Commit:
Add missing docstrings.

Commit:
Add resampling capability to drive cycle

Commit:
Force dist and elev to recalculate at end of maneuver.

Commit:
Check coasting on grade

michael-okeefe commented:
@calbaker @jakeholden 
This update now includes a re-rendezvous to the reference trace if "Correct" is chosen for trace miss option. There is a demo file for the trace miss and also I have added a demo file for showing coasting with grade. Unfortunately, there is something slightly off with coasting on grade but it is probably close enough to move forward with -- there is a new demo in `demo_cavs.py` that showcases running coasting on grade so you can see for yourself. Please let me know if there are any questions. I also left a more extensive comment here:
https://github.nrel.gov/MBAP/fastsim/issues/380

Commit:
Add TODO item

calbaker commented:
@michael-okeefe , is this one ready for review and merging?  If yes, I'll give it a deep dive.

michael-okeefe commented:
> @michael-okeefe , is this one ready for review and merging? If yes, I'll give it a deep dive.

@calbaker Hi Chad, yes, this should be ready for review now (with the few caveats mentioned). Let me know if you have any questions. Thank you!

Commit:
Merge branch 'fastsim-3' of github.com:NREL/fastsim into f3/acc-features
this merge broke a bunch of `ninterp` stuff which needs to be checked by either Kyle or Robin

robinsteuteville committed:
fixing ninterp errors

Commit:
worked out python problems
